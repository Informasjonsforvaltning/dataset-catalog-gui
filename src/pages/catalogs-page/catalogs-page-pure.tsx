import React from 'react';
import { CardGroup } from 'reactstrap';
import _ from 'lodash';
import localization from '../../services/localization';
import { Catalog } from './catalogs/catalog.component';
import { getTranslateText } from '../../services/translateText';
import { selectorForCatalogDatasetsFromDatasetsState } from '../../redux/modules/datasets';
import { getAPIItemsCount } from '../../redux/modules/apis';
import './catalogs-page.scss';
import { authService } from '../../services/auth/auth-service';

interface Props {
  catalogItems: any[];
  isFetchingCatalogs: boolean;
  datasetsState: any;
  apis: any;
  fetchCatalogsIfNeeded: () => void;
  fetchDatasetsIfNeeded: () => void;
  fetchApisIfNeeded: () => void;
}

export const CatalogsPagePure = ({
  isFetchingCatalogs = false,
  fetchCatalogsIfNeeded,
  catalogItems,
  datasetsState,
  apis,
  fetchDatasetsIfNeeded,
  fetchApisIfNeeded
}: Props) => {
  fetchCatalogsIfNeeded && fetchCatalogsIfNeeded();

  if (isFetchingCatalogs) {
    return null;
  }

  return (
    <div className="container">
      {catalogItems &&
        catalogItems.map(catalog => (
          <div key={_.get(catalog, 'id')} className="row mb-2 mb-md-5">
            <div className="col-12">
              <div>
                <h2 className="fdk-text-strong mb-4">
                  {getTranslateText(
                    _.get(catalog, ['publisher', 'prefLabel'])
                  ) || _.get(catalog, ['publisher', 'name'], '')}
                </h2>
              </div>
              <CardGroup>
                {datasetsState && (
                  <Catalog
                    key={`datasets-${catalog.id}`}
                    catalogId={catalog.id}
                    type="datasets"
                    fetchItems={fetchDatasetsIfNeeded}
                    itemsCount={
                      Object.keys(
                        selectorForCatalogDatasetsFromDatasetsState(catalog.id)(
                          datasetsState
                        )
                      ).length
                    }
                  />
                )}
                {apis && (
                  <Catalog
                    key={`apis-${catalog.id}`}
                    catalogId={catalog.id}
                    fetchItems={fetchApisIfNeeded}
                    type="apis"
                    itemsCount={getAPIItemsCount(apis, catalog.id)}
                    isReadOnly={
                      !authService.hasOrganizationWritePermission(catalog.id)
                    }
                  />
                )}
                <Catalog
                  key={`concepts-${catalog.id}`}
                  catalogId={catalog.id}
                  type="concepts"
                />
                <Catalog
                  key={`protocol-${catalog.id}`}
                  catalogId={catalog.id}
                  type="protocol"
                  isReadOnly={
                    !authService.hasOrganizationWritePermission(catalog.id)
                  }
                />
              </CardGroup>
            </div>
          </div>
        ))}
      {!catalogItems && (
        <div className="row mb-2 mb-md-5">
          <div id="no-catalogs">
            <h1 className="fdk-text-strong">
              {localization.catalogs.missingCatalogs.title}
            </h1>
            <div className="mt-2 mb-2">
              {localization.catalogs.missingCatalogs.ingress}
            </div>
            <div className="fdk-text-size-small">
              <strong>
                {localization.catalogs.missingCatalogs.accessTitle}
              </strong>
              <p>
                <a href="https://fellesdatakatalog.brreg.no/about-registration">
                  {localization.catalogs.missingCatalogs.accessText}
                  <i className="fa fa-external-link fdk-fa-right" />
                </a>
              </p>
            </div>
          </div>
        </div>
      )}
    </div>
  );
};
