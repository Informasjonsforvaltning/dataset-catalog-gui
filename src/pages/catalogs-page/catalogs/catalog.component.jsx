import React from 'react';
import PropTypes from 'prop-types';
import memoize from 'lodash/memoize';

import { resolve } from 'react-resolver';
import { getConfig } from '../../../config';
import { CatalogItem } from './catalog-item/catalog-item.component';
import { getConceptCount } from '../../../services/api/concept-registration-api/host';
import { getRecordsCount } from '../../../services/api/records-registration-api/host';

export const CatalogPure = props => {
  const { catalogId, type, fetchItems, itemsCount, isReadOnly } = props;
  fetchItems(catalogId);

  const getLinkUri = () => {
    switch (type) {
      case 'concepts': {
        return `${getConfig().conceptRegistrationHost}/${catalogId}`;
      }
      case 'protocol': {
        return `${getConfig().recordsOfProcessingActivitiesHost}/${catalogId}`;
      }
      default:
        return `/catalogs/${catalogId}/${type}`;
    }
  };

  const linkUri = getLinkUri();

  return (
    <CatalogItem
      linkUri={linkUri}
      key={catalogId}
      publisherId={catalogId}
      type={type}
      itemsCount={itemsCount}
      isReadOnly={isReadOnly}
    />
  );
};

CatalogPure.defaultProps = {
  catalogId: null,
  fetchItems: () => {},
  itemsCount: null,
  isReadOnly: false
};

CatalogPure.propTypes = {
  catalogId: PropTypes.string,
  type: PropTypes.string.isRequired,
  fetchItems: PropTypes.func,
  itemsCount: PropTypes.number,
  isReadOnly: PropTypes.bool
};

const memoizedGetConceptCount = memoize(getConceptCount);
const memoizedGetRecordsCount = memoize(getRecordsCount);

const mapProps = {
  itemsCount: ({ type, catalogId, itemsCount }) => {
    switch (type) {
      case 'concepts': {
        return memoizedGetConceptCount(catalogId);
      }
      case 'protocol': {
        return memoizedGetRecordsCount(catalogId);
      }
      default:
        return itemsCount;
    }
  }
};

export const Catalog = resolve(mapProps)(CatalogPure);
