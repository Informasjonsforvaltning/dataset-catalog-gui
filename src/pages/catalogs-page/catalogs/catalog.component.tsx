import React, { memo, FC } from 'react';
import { compose } from 'redux';
import memoize from 'lodash/memoize';
import { resolve } from 'react-resolver';

import env from '../../../env';

import { getConceptCount } from '../../../services/api/concept-registration-api/host';
import { getRecordsCount } from '../../../services/api/records-registration-api/host';
import { getDataServicesCount } from '../../../services/api/dataservice-catalog/host';

import CatalogItem from './catalog-item/catalog-item.component';

const {
  DATASERVICE_CATALOG_BASE_URI,
  CONCEPT_REGISTRATION_HOST,
  RECORDS_OF_PROCESSING_ACTIVITIES_GUI_BASE_URI
} = env;

interface ExternalProps {
  catalogId: string;
  type: string;
  fetchItems?: (id: string) => void;
  itemsCount?: number;
  isReadOnly?: boolean;
  disabled: boolean;
}

interface Props extends ExternalProps {}

const Catalog: FC<Props> = ({
  catalogId,
  type,
  fetchItems,
  itemsCount,
  isReadOnly,
  disabled
}) => {
  fetchItems?.(catalogId);

  const getLinkUri = () => {
    switch (type) {
      case 'dataservices': {
        return `${DATASERVICE_CATALOG_BASE_URI}/${catalogId}`;
      }
      case 'concepts': {
        return `${CONCEPT_REGISTRATION_HOST}/${catalogId}`;
      }
      case 'protocol': {
        return `${RECORDS_OF_PROCESSING_ACTIVITIES_GUI_BASE_URI}/${catalogId}`;
      }
      default:
        return `/catalogs/${catalogId}/${type}`;
    }
  };

  const linkUri = getLinkUri();

  return (
    <CatalogItem
      linkUri={linkUri}
      key={catalogId}
      type={type}
      itemsCount={itemsCount}
      isReadOnly={isReadOnly}
      disabled={disabled}
    />
  );
};

const memoizedGetDataServicesCount = memoize(getDataServicesCount);
const memoizedGetConceptCount = memoize(getConceptCount);
const memoizedGetRecordsCount = memoize(getRecordsCount);

const mapProps = {
  itemsCount: ({ type, catalogId, itemsCount }: any) => {
    switch (type) {
      case 'dataservices': {
        return memoizedGetDataServicesCount(catalogId);
      }
      case 'concepts': {
        return memoizedGetConceptCount(catalogId);
      }
      case 'protocol': {
        return memoizedGetRecordsCount(catalogId);
      }
      default:
        return itemsCount;
    }
  }
};

export default compose<FC<ExternalProps>>(memo, resolve(mapProps))(Catalog);
