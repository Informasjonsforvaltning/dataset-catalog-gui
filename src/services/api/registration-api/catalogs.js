export const catalogsPath = `/catalogs`;
export const catalogPath = catalogId => `${catalogsPath}/${catalogId}`;
