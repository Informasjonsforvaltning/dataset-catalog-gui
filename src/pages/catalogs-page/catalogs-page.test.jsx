import React from 'react';
import { shallow } from 'enzyme';
import { CatalogsPagePure } from './catalogs-page-pure';
import catalogs from '../../../test/fixtures/catalogs';

let defaultProps;
let wrapper;
let dispatch;

beforeEach(() => {
  dispatch = jest.fn();
  defaultProps = {
    dispatch,
    catalogItems: catalogs.catalogItems,
    isFetchingCatalogs: false
  };
  wrapper = shallow(<CatalogsPagePure {...defaultProps} />);
});

test('should render CatalogsPagePure correctly', () => {
  expect(wrapper).toMatchSnapshot();
});
