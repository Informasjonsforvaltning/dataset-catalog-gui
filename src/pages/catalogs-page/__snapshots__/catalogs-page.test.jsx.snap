// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CatalogsPagePure correctly 1`] = `
<div
  className="container"
>
  <div
    className="row mb-2 mb-md-5"
    key="910244132"
  >
    <div
      className="col-12"
    >
      <div>
        <h2
          className="fdk-text-strong mb-4"
        >
          RAMSUND OG ROGNAN REVISJON
        </h2>
      </div>
      <CardGroup
        tag="div"
      >
        <ItemsCountResolver
          catalogId="910244132"
          key="concepts-910244132"
          type="concepts"
        />
        <ItemsCountResolver
          catalogId="910244132"
          isReadOnly={true}
          key="protocol-910244132"
          type="protocol"
        />
      </CardGroup>
    </div>
  </div>
</div>
`;
